### ISSUE TITLE
out of memory crash, built from scratch on Debian and Ubuntu

### ISSUE THREAD

**Issue by hvqzao** [npm/npm#14565]
#### I'm opening this issue because:

  - [x] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

I've build nodejs 6.9.1 on Debian Rolling and Ubuntu 16.10. On both npm search anything crashes. I noticed people are complaining about such crashes for months on Twitter. I'm using LTS version. I'm testing it on 4 GB / 8 GB VMware VM's with 2 GB of swap.

#### How can the CLI team reproduce the problem?

compile newest nodejs in virtual env (I'm using current "nodeenv" python package to set it up) and try to use npm, e.g. (here I'm compiling python and nodejs):

```sh
mkdir a
cd a
git clone https://github.com/hvqzao/ve
cd ve
sudo ./ve -r   # installs requirements
./ve -p 3.5.2    # builds python
[wait]
<press Enter>           # to confirm ./configure
[wait]
./ve -N 3.5.2 6.9.1   # build nodejs
. ./ve node-6.9.1       # use nodejs
npm search anything
deact                          #  deactivates virtual environment
```

result:

```
(node-6.9.1) qzao@ubuntu:~/a/ve$ npm search omg
npm WARN Building the local index for the first time, please be patient
[..................] \ : WARN Building the local index for the first time, please be patient
<--- Last few GCs --->

   45507 ms: Mark-sweep 1053.0 (1375.4) -> 1027.1 (1354.1) MB, 1473.1 / 0.0 ms [allocation failure] [GC in old space requested].
   47008 ms: Mark-sweep 1027.1 (1354.1) -> 1027.0 (1314.1) MB, 1501.2 / 0.0 ms [last resort gc].
   48481 ms: Mark-sweep 1027.0 (1314.1) -> 1026.9 (1307.1) MB, 1473.3 / 0.0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x3fa0a95cfb51 <JS Object>
    1: fromString(aka fromString) [buffer.js:199] [pc=0x204120ed5790] (this=0x3fa0a9504381 <undefined>,string=0x1239d3054549 <Very long string[214763301]>,encoding=0x3fa0a95dd361 <String[4]: utf8>)
    2: from [buffer.js:99] [pc=0x204120e407a2] (this=0x1063d0aa8259 <JS Function Buffer (SharedFunctionInfo 0x2e4e71ef71d1)>,value=0x1239d3054549 <Very long string[214763301]>,encodingOrOffset=0x3fa0...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [npm]
 2: 0x1098b2c [npm]
 3: v8::Utils::ReportApiFailure(char const*, char const*) [npm]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [npm]
 5: v8::internal::Factory::NewRawTwoByteString(int, v8::internal::PretenureFlag) [npm]
 6: v8::internal::String::SlowFlatten(v8::internal::Handle<v8::internal::ConsString>, v8::internal::PretenureFlag) [npm]
 7: v8::internal::String::Flatten(v8::internal::Handle<v8::internal::String>, v8::internal::PretenureFlag) [npm]
 8: v8::String::WriteUtf8(char*, int, int*, int) const [npm]
 9: node::StringBytes::Write(v8::Isolate*, char*, unsigned long, v8::Local<v8::Value>, node::encoding, int*) [npm]
10: node::Buffer::New(v8::Isolate*, v8::Local<v8::String>, node::encoding) [npm]
11: node::Buffer::CreateFromString(v8::FunctionCallbackInfo<v8::Value> const&) [npm]
12: v8::internal::FunctionCallbackArguments::Call(void (*)(v8::FunctionCallbackInfo<v8::Value> const&)) [npm]
13: 0x9e6f22 [npm]
14: 0x9e759e [npm]
15: 0x204120a092a7
Aborted (core dumped)
```

again:

```
(node-6.9.1) qzao@ubuntu:~/a/ve$ npm search asd
npm ERR! Linux 4.8.0-26-generic
npm ERR! argv "/home/qzao/a/ve/build/node-6.9.1/bin/node" "/home/qzao/a/ve/build/node-6.9.1/bin/npm" "search" "asd"
npm ERR! node v6.9.1
npm ERR! npm  v3.10.8

npm ERR! Cannot read property '_updated' of undefined
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>
npm WARN Building the local index for the first time, please be patient

<--- Last few GCs --->

  100469 ms: Mark-sweep 1052.8 (1373.4) -> 1026.9 (1354.1) MB, 2610.5 / 0.0 ms [allocation failure] [GC in old space requested].
  103686 ms: Mark-sweep 1026.9 (1354.1) -> 1026.8 (1314.1) MB, 3216.4 / 0.0 ms [last resort gc].
  106734 ms: Mark-sweep 1026.8 (1314.1) -> 1026.8 (1308.1) MB, 3048.4 / 0.0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x1730baecfb51 <JS Object>
    1: fromString(aka fromString) [buffer.js:199] [pc=0x353c3a5be910] (this=0x1730bae04381 <undefined>,string=0xd8144b4d519 <Very long string[214763301]>,encoding=0x1730baedd361 <String[4]: utf8>)
    2: from [buffer.js:99] [pc=0x353c3a54c7c2] (this=0x869bbda1fd1 <JS Function Buffer (SharedFunctionInfo 0x2874331f71d1)>,value=0xd8144b4d519 <Very long string[214763301]>,encodingOrOffset=0x1730bae...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [npm]
 2: 0x1098b2c [npm]
 3: v8::Utils::ReportApiFailure(char const*, char const*) [npm]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [npm]
 5: v8::internal::Factory::NewRawTwoByteString(int, v8::internal::PretenureFlag) [npm]
 6: v8::internal::String::SlowFlatten(v8::internal::Handle<v8::internal::ConsString>, v8::internal::PretenureFlag) [npm]
 7: v8::internal::String::Flatten(v8::internal::Handle<v8::internal::String>, v8::internal::PretenureFlag) [npm]
 8: v8::String::WriteUtf8(char*, int, int*, int) const [npm]
 9: node::StringBytes::Write(v8::Isolate*, char*, unsigned long, v8::Local<v8::Value>, node::encoding, int*) [npm]
10: node::Buffer::New(v8::Isolate*, v8::Local<v8::String>, node::encoding) [npm]
11: node::Buffer::CreateFromString(v8::FunctionCallbackInfo<v8::Value> const&) [npm]
12: v8::internal::FunctionCallbackArguments::Call(void (*)(v8::FunctionCallbackInfo<v8::Value> const&)) [npm]
13: 0x9e6f22 [npm]
14: 0x9e759e [npm]
15: 0x353c3a0092a7
Aborted (core dumped)

```

### supporting information:

 - `npm -v` prints:

3.10.8

 - `node -v` prints:

v6.9.1

 - `npm config get registry` prints:

https://registry.npmjs.org/

 - Windows, OS X/macOS, or Linux?:

Linux

 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by legodude17** [comment]
This is a known issue with anything below `npm@4` please update to `npm@4` if you want to resolve this.

**Comment by othiym23** [comment]
Closing as a duplicate of #12619, fixed by #13746 – update to `npm@next` to get a working search implementation. Thanks for your time!
